src/orion/algo/gridsearch.py

```diff
@@ -0,0 +1,225 @@
+
+        discrete_grid.append(int_v)
+
+    return discrete_grid
+
+
+def real_grid(dim, num):
+    """Build real grid"""
+    if dim.prior_name.endswith("reciprocal"):
+        a, b = dim.interval()
+        return list(_log_grid(a, b, num))
+    elif dim.prior_name.endswith("uniform"):
+        a, b = dim.interval()
+        return list(_lin_grid(a, b, num))
+    else:
+        raise TypeError(
+            "Grid Search only supports `loguniform`, `uniform` and `choices`: "
+            "`{}`".format(dim.prior_name)
+        )
+
+
+def _log_grid(a, b, num):
+    return numpy.exp(_lin_grid(numpy.log(a), numpy.log(b), num))
+
+
+def _lin_grid(a, b, num):
+    return numpy.linspace(a, b, num=num)
+
+
+class GridSearch(BaseAlgorithm):
+    """Grid Search algorithm
+
+    Parameters
+    ----------
+    n_values: int or dict
+        Number of points for each dimensions, or dictionary specifying number of points for each
+        dimension independently (name, n_values). For categorical dimensions, n_values will not be
+        used, and all categories will be used to build the grid.
+    """
+
+    requires_type = None
+    requires_dist = None
+    requires_shape = "flattened"
+
+    def __init__(self, space, n_values=100, seed=None):
+        super(GridSearch, self).__init__(space, n_values=n_values, seed=seed)
+        self.n = 0
+        self.grid = None
+
+    def _initialize(self):
+        """Initialize the grid once the space is transformed"""
+        n_values = self.n_values
+        if not isinstance(n_values, dict):
+            n_values = {name: self.n_values for name in self.space.keys()}
+
+        self.grid = self.build_grid(
+            self.space, n_values, getattr(self, "max_trials", 10000)
+        )
+
+    @staticmethod
+    def build_grid(space, n_values, max_trials=10000):
+        """Build a grid of trials
+
+        Parameters
+        ----------
+        n_values: int or dict
+            Number of points for each dimensions, or dictionary specifying number of points for each
+            dimension independently (name, n_values). For categorical dimensions, n_values will not be
+            used, and all categories will be used to build the grid.
+        max_trials: int
+            Maximum number of trials for the grid. If n_values lead to more trials than max_trials,
+            the n_values will be adjusted down. Will raise ValueError if it is impossible to build
+            a grid smaller than max_trials (for instance if choices are too large).
+
+        """
+        adjust = 0
+        n_trials = float("inf")
+
+        while n_trials > max_trials:
+            coordinates = []
+            capped_values = []
+            for name, dim in space.items():
+                capped_value = max(n_values[name] - adjust, 1)
+                capped_values.append(capped_value)
+                coordinates.append(list(grid(dim, capped_value)))
+
+            if all(value <= 1 for value in capped_values):
```

Comment:
looks like capped_value for all dimension will always be the same at this place.  So we want `raise` if any dimension `n_values` is adjusted to 1 or all the dimension is adjust?

Reply:
`n_values[name]` can vary from one dimension to another. We want to keep adjusting until grid is >= max_trials or until there is no more adjustment possible (all dims have only 1 point each). If this is still larger than max_trials, raise ValueError.

Reply:
`n_values[name]` can vary from one dimension to another. We want to keep adjusting until grid is >= max_trials or until there is no more adjustment possible (all dims have only 1 point each). If this is still larger than max_trials, raise ValueError.

Reply:
How about a space like below. if `max_trials` is 10, when `n_values` = 2 for all dims, the grid size is 16, but in a middle of adjustment such as `n_values[dim1]=1, n_values[dim2]=2, n_values[dim3]=2, n_values[dim4]=2`, the grid size is 8, which looks like a valid gride.

```
    dim1 = Real("dim1", "uniform", 0, 1)
    dim2 = Integer("dim2", "uniform", 0, 10)
    dim3 = Categorical("dim3", "ab")
    dim4 = Categorical("dim4", "cd")
```

Reply:
Yeah we could do the adjustment in a more efficient way, adjusting by 1 for all incrementally is very basic. Lazy me is tempted to say current version is good enough however. :) Do you think leaving this improvement for later would be harmful for the usability of the algorithm? 

Or perhaps I could just add an `if` test in the for-loop to break earlier if we reach n_trials < max_trials.

Reply:
I think current implementation is okay, just want to confirm this is what we want to do. :)